src/quantization/scalar_quantization_trainer.cpp

```diff
@@ -0,0 +1,100 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "scalar_quantization_trainer.h"
+
+#include <random>
+#include <unordered_set>
+
+#include "simd/normalize.h"
+
+namespace vsag {
+
+ScalarQuantizationTrainer::ScalarQuantizationTrainer(int32_t dim, int bits)
+    : dim_(dim), bits_(bits) {
+}
+
+void
+ScalarQuantizationTrainer::Train(const float* data,
+                                 uint64_t count,
+                                 float* upper_bound,
+                                 float* lower_bound,
+                                 bool need_normalize,
+                                 ScalarQuantizationTrainer::SQTrainMode mode) {
+    std::vector<float> sample_datas;
+    auto sample_count = this->sample_train_data(data, count, sample_datas, need_normalize);
+    if (mode == CLASSIC) {
```

Comment:
add unsupported mode check and print warn log ?

Reply:
this is inner mode, I just want to extend it in future

Reply:
Ok, remember to add check